flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTestHarness.java

```diff
@@ -231,15 +230,20 @@ public Thread invoke() throws Exception {
 	 *
 	 */
 	public Thread invoke(StreamMockEnvironment mockEnv) throws Exception {
+		checkState(this.mockEnv == null);
+		checkState(this.taskThread == null);
 		this.mockEnv = checkNotNull(mockEnv);
 
 		initializeInputs();
 		initializeOutput();
 
-		this.task = taskFactory.apply(mockEnv);
-
-		taskThread = new TaskThread(task);
+		taskThread = new TaskThread(() -> taskFactory.apply(mockEnv));
 		taskThread.start();
+		// Wait until the task is set
+		while (taskThread.task == null) {
+			Thread.sleep(10L);
```

Comment:
Nit: Why not use a latch like `OneShotLatch` here to wait-block until the field is set and the thread then signals?

Reply:
Can be done as you propose by using a `OneShotLatch`, but this would add a dedicated field in `TaskThread` in that case.